File: src/main/java/tn/esprit/entities/reservation_local.java
package tn.esprit.entities;

import tn.esprit.enums.StatutReservation;
import tn.esprit.enums.MotifReservation;
import java.util.Date;
import java.util.List;

public class reservation_local {
    private int id_reservation;
    private int id_utilisateur;
    private int id_local;
    private int salle_id;          // FK → salle (0 = aucune salle réservée)
    private Date date_reservation;
    private Date heure_debut_reservation;
    private Date heure_fin_reservation;
    private StatutReservation status_reservation;
    private MotifReservation motif_reservation;
    private int prix_reservation;
    private String nom_cl;
    private String prenom_cl;

    // Constructeur vide
    public reservation_local() {
    }

    // Constructeur avec tous les paramètres (sans salle)
    public reservation_local(int id_reservation, int id_utilisateur, int id_local, Date date_reservation,
                             Date heure_debut_reservation, Date heure_fin_reservation, StatutReservation status_reservation,
                             MotifReservation motif_reservation, int prix_reservation, String nom_cl, String prenom_cl) {
        this.id_reservation = id_reservation;
        this.id_utilisateur = id_utilisateur;
        this.id_local = id_local;
        this.salle_id = 0;
        this.date_reservation = date_reservation;
        this.heure_debut_reservation = heure_debut_reservation;
        this.heure_fin_reservation = heure_fin_reservation;
        this.status_reservation = status_reservation;
        this.motif_reservation = motif_reservation;
        this.prix_reservation = prix_reservation;
        this.nom_cl = nom_cl;
        this.prenom_cl = prenom_cl;
    }

    // Constructeur avec salle_id
    public reservation_local(int id_reservation, int id_utilisateur, int id_local, int salle_id,
                             Date date_reservation, Date heure_debut_reservation, Date heure_fin_reservation,
                             StatutReservation status_reservation, MotifReservation motif_reservation,
                             int prix_reservation, String nom_cl, String prenom_cl) {
        this.id_reservation = id_reservation;
        this.id_utilisateur = id_utilisateur;
        this.id_local = id_local;
        this.salle_id = salle_id;
        this.date_reservation = date_reservation;
        this.heure_debut_reservation = heure_debut_reservation;
        this.heure_fin_reservation = heure_fin_reservation;
        this.status_reservation = status_reservation;
        this.motif_reservation = motif_reservation;
        this.prix_reservation = prix_reservation;
        this.nom_cl = nom_cl;
        this.prenom_cl = prenom_cl;
    }



    // Getters et Setters
    public int getId_reservation() {
        return id_reservation;
    }

    public void setId_reservation(int id_reservation) {
        this.id_reservation = id_reservation;
    }

    public int getId_utilisateur() {
        return id_utilisateur;
    }

    public void setId_utilisateur(int id_utilisateur) {
        this.id_utilisateur = id_utilisateur;
    }

    public int getId_local() {
        return id_local;
    }

    public void setId_local(int id_local) {
        this.id_local = id_local;
    }

    public int getSalle_id() {
        return salle_id;
    }

    public void setSalle_id(int salle_id) {
        this.salle_id = salle_id;
    }

    public Date getDate_reservation() {
        return date_reservation;
    }

    public void setDate_reservation(Date date_reservation) {
        this.date_reservation = date_reservation;
    }

    public Date getHeure_debut_reservation() {
        return heure_debut_reservation;
    }

    public void setHeure_debut_reservation(Date heure_debut_reservation) {
        this.heure_debut_reservation = heure_debut_reservation;
    }

    public Date getHeure_fin_reservation() {
        return heure_fin_reservation;
    }

    public void setHeure_fin_reservation(Date heure_fin_reservation) {
        this.heure_fin_reservation = heure_fin_reservation;
    }

    public StatutReservation getStatus_reservation() {
        return status_reservation;
    }

    public void setStatus_reservation(StatutReservation status_reservation) {
        this.status_reservation = status_reservation;
    }

    public MotifReservation getMotif_reservation() {
        return motif_reservation;
    }

    public void setMotif_reservation(MotifReservation motif_reservation) {
        this.motif_reservation = motif_reservation;
    }

    public int getPrix_reservation() {
        return prix_reservation;
    }

    public void setPrix_reservation(int prix_reservation) {
        this.prix_reservation = prix_reservation;
    }

    public String getNom_cl() {
        return nom_cl;
    }

    public void setNom_cl(String nom_cl) {
        this.nom_cl = nom_cl;
    }

    public String getPrenom_cl() {
        return prenom_cl;
    }

    public void setPrenom_cl(String prenom_cl) {
        this.prenom_cl = prenom_cl;
    }

    @Override
    public String toString() {
        return "reservation{" +
                "id_reservation=" + id_reservation +
                ", id_utilisateur=" + id_utilisateur +
                ", id_local=" + id_local +
                ", salle_id=" + salle_id +
                ", date_reservation=" + date_reservation +
                ", heure_debut_reservation=" + heure_debut_reservation +
                ", heure_fin_reservation=" + heure_fin_reservation +
                ", status_reservation=" + status_reservation +
                ", motif_reservation=" + motif_reservation +
                ", prix_reservation=" + prix_reservation +
                ", nom_cl='" + nom_cl + '\'' +
                ", prenom_cl='" + prenom_cl + '\'' +
                '}';
    }

}

File: src/Entity/ReservationLocal.php
<?php

namespace App\Entity;

use App\Repository\ReservationLocalRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;
use App\Entity\Salle;
use App\Entity\LocalsPsychiatrie;
use App\Entity\ReservationHistorique;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;

#[ORM\Entity(repositoryClass: ReservationLocalRepository::class)]
#[ORM\Table(name: 'reservation_local')]
class ReservationLocal
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'id_reservation', type: Types::INTEGER)]
    /** @phpstan-ignore property.unusedType */
    private ?int $id = null;

    #[ORM\ManyToOne(targetEntity: Salle::class)]
    #[ORM\JoinColumn(name: 'salle_id', referencedColumnName: 'id_salle', nullable: false, onDelete: "CASCADE")]
    private ?Salle $salle = null;

    #[ORM\Column(name: 'id_utilisateur', type: Types::INTEGER, nullable: true)]
    private ?int $idUtilisateur = null;

    #[ORM\ManyToOne(targetEntity: LocalsPsychiatrie::class)]
    #[ORM\JoinColumn(name: 'local_id', referencedColumnName: 'id_local', nullable: true)]
    private ?LocalsPsychiatrie $local = null;

    #[ORM\Column(name: 'date_reservation', type: Types::DATE_MUTABLE, nullable: true)]
    #[Assert\NotBlank(message: 'La date de réservation est obligatoire.')]
    private ?\DateTimeInterface $dateReservation = null;

    #[ORM\Column(name: 'heure_debut_reservation', type: Types::TIME_MUTABLE, nullable: true)]
    #[Assert\NotBlank(message: "L'heure de début est obligatoire.")]
    private ?\DateTimeInterface $heureDebutReservation = null;

    #[ORM\Column(name: 'heure_fin_reservation', type: Types::TIME_MUTABLE, nullable: true)]
    #[Assert\NotBlank(message: "L'heure de fin est obligatoire.")]
    private ?\DateTimeInterface $heureFinReservation = null;

    #[ORM\Column(name: 'status_reservation', type: Types::STRING, length: 50, nullable: true)]
    private ?string $statusReservation = 'En attente';

    #[ORM\Column(name: 'motif_reservation', type: Types::TEXT, nullable: true)]
    private ?string $motifReservation = null;

    #[ORM\Column(name: 'prix_reservation', type: Types::DECIMAL, precision: 10, scale: 2, nullable: true)]
    private ?string $prixReservation = null;

    #[ORM\Column(name: 'nom_cl', type: Types::STRING, length: 100, nullable: true)]
    #[Assert\NotBlank(message: 'Le nom du client est obligatoire.')]
    private ?string $nomCl = null;

    #[ORM\Column(name: 'prenom_cl', type: Types::STRING, length: 100, nullable: true)]
    #[Assert\NotBlank(message: 'Le prénom du client est obligatoire.')]
    private ?string $prenomCl = null;

    // 🔥 RELATION MANQUANTE AJOUTÉE
    /** @var Collection<int, ReservationHistorique> */
    #[ORM\OneToMany(mappedBy: 'reservation', targetEntity: ReservationHistorique::class)]
    private Collection $historiques;

    // ── Constructor ─────────────────────────────
    public function __construct()
    {
        $this->historiques = new ArrayCollection();
    }

    // ── Getters / Setters ───────────────────────

    public function getIdReservation(): ?int
    {
        return $this->id;
    }

    public function getSalle(): ?Salle
    {
        return $this->salle;
    }

    public function setSalle(?Salle $salle): static
    {
        $this->salle = $salle;
        return $this;
    }

    public function getIdUtilisateur(): ?int
    {
        return $this->idUtilisateur;
    }

    public function setIdUtilisateur(?int $idUtilisateur): static
    {
        $this->idUtilisateur = $idUtilisateur;
        return $this;
    }

    public function getLocal(): ?LocalsPsychiatrie
    {
        return $this->local;
    }

    public function setLocal(?LocalsPsychiatrie $local): static
    {
        $this->local = $local;
        return $this;
    }

    public function getDateReservation(): ?\DateTimeInterface
    {
        return $this->dateReservation;
    }

    public function setDateReservation(?\DateTimeInterface $dateReservation): static
    {
        $this->dateReservation = $dateReservation;
        return $this;
    }

    public function getHeureDebutReservation(): ?\DateTimeInterface
    {
        return $this->heureDebutReservation;
    }

    public function setHeureDebutReservation(?\DateTimeInterface $heureDebutReservation): static
    {
        $this->heureDebutReservation = $heureDebutReservation;
        return $this;
    }

    public function getHeureFinReservation(): ?\DateTimeInterface
    {
        return $this->heureFinReservation;
    }

    public function setHeureFinReservation(?\DateTimeInterface $heureFinReservation): static
    {
        $this->heureFinReservation = $heureFinReservation;
        return $this;
    }

    public function getStatusReservation(): ?string
    {
        return $this->statusReservation;
    }

    public function setStatusReservation(?string $statusReservation): static
    {
        $this->statusReservation = $statusReservation;
        return $this;
    }

    public function getMotifReservation(): ?string
    {
        return $this->motifReservation;
    }

    public function setMotifReservation(?string $motifReservation): static
    {
        $this->motifReservation = $motifReservation;
        return $this;
    }

    public function getPrixReservation(): ?string
    {
        return $this->prixReservation;
    }

    public function setPrixReservation(?string $prixReservation): static
    {
        $this->prixReservation = $prixReservation;
        return $this;
    }

    public function getNomCl(): ?string
    {
        return $this->nomCl;
    }

    public function setNomCl(?string $nomCl): static
    {
        $this->nomCl = $nomCl;
        return $this;
    }

    public function getPrenomCl(): ?string
    {
        return $this->prenomCl;
    }

    public function setPrenomCl(?string $prenomCl): static
    {
        $this->prenomCl = $prenomCl;
        return $this;
    }

    // ── Historiques ─────────────────────────────

    /** @return Collection<int, ReservationHistorique> */
    public function getHistoriques(): Collection
    {
        return $this->historiques;
    }
}